Lateral profile

MarekOchocki opened the pull request:
#10

MarekOchocki committed:
Implement methods in lateral class

- penumbra (left and right)
- field_ratio
- symmetry
- flatness_50
- flatness_90
- asymmetry
- normalize
- absolute_y

MarekOchocki committed:
Add comments explaining methods

- docstring explaining penumbra
- docstring explaining absolute_y
- imports sorted

MarekOchocki committed:
Replace constant integers with floats

MarekOchocki committed:
Merge absolute_y with normalize

You have to call absolute_y before calling normalize anyway.
If you call these in opposite way you'd have to call normalize again.

MarekOchocki committed:
Cosmetic changes to fit PEP8 style

MarekOchocki committed:
Cosmetic changes part 2

MarekOchocki committed:
Adding allow_cast in normalize

- allow_cast in normalize
- fixed penumbras

MarekOchocki committed:
Merge branch 'master' into lateral_profile

grzanka commented:
@MarekOchocki PEP8 again

MarekOchocki committed:
Cosmetic changes to fit PEP8

MarekOchocki committed:
Add handling with corrupted data

grzanka commented:
@MarekOchocki is it ready ?

MarekOchocki commented:
Not yet. I'm doing final changes

MarekOchocki committed:
Fix raising exception in normalize()

MarekOchocki committed:
Change implementations to be more accurate

- Fix ignoring part of the profile in the middle in asymmetry()
- Add searching for max/min value on borders in flatness_90 and flatness_50

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -23,48 +23,79 @@ def penumbra_left(self):
+        """
         d = (self.penumbra_left() + self.penumbra_right()) / 2
-        if d == np.nan:
+        if d != d:
```
Add a comment why such interesting comparison is being made. Is it a test for corrupted number ?
If yes, the please explain it.

MarekOchocki replied:
standard for representing floating-point numbers requires that NaN != NaN is always true. d != d is the same comparison as np.isnan(d)

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -23,48 +23,79 @@ def penumbra_left(self):
-        p_max = np.max(self.y[np.logical_and(self.x <= right, self.x >= left)])
-        p_min = np.min(self.y[np.logical_and(self.x <= right, self.x >= left)])
+        p_max = np.max(
+            np.append(self.y[np.logical_and(self.x <= right, self.x >= left)], [self.y_at_x(right), self.y_at_x(left)]))
```
Line gets really long, can you split it ?

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -23,48 +23,79 @@ def penumbra_left(self):
+            area_right += np.trapz(np.append(self.y_at_x(0), self.y[right_index_arr[0]]),
+                                  np.append([.0], self.x[right_index_arr[0]]))
+
+        return float(((area_left - area_right) / (area_left + area_right)) * 100.0)
```
Why conversion to float is needed ?

MarekOchocki replied:
If we want to check returned value in tests, returned value has to be some type that has round method.  Previous version was returning DataSet instead of float. We can convert DataSet to float in LateralProfile or tests. I decided to do it in LateralProfile

grzanka replied:
OK, so I propose to explain it in some comments in the code (code should be easy to read for future developers).

Does it mean that `area_left` and `area_right` are also numpy DataSets and it propagates to return value ? If yes, then please put this info in comments.
numpy documentation of trapz says it returns floats:
https://docs.scipy.org/doc/numpy/reference/generated/numpy.trapz.html

You may also consider introducing another variable called result (of DataSet type) and do `return float(result)`

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -23,48 +23,79 @@ def penumbra_left(self):
 
     def asymmetry(self):
+        """
+        In case of corrupted data returns nan
```
Does it return nan or np.nan ?

MarekOchocki replied:
Doesn't matter, because type(np.nan) is equal type(float('nan'))

MarekOchocki committed:
Extend comments

Additionally, change checking for NaN from d != d to np.isnan(d) to make it less confusing

grzanka commented on ibodata/lateral_profile.py:
```diff
@@ -6,56 +6,107 @@
+    def penumbra_right(self):
         """
-        In case of good data returns positive number
+        In case of good data returns float
```
It is always float ? What if profile contains integer numbers, like:
`X = [0,1,2,3]`  and `Y=[0,1,10,1]`

Better use floating-point number than float

MarekOchocki committed:
Change code to fit PEP8
